### PR TITLE
Issue #9455: Introduce Gecko Credit Cards and Addresses Storage Delegate

### LINKED ISSUE #9455: Gecko credit card storage delegates
Larger issue: #9452

Once we have https://github.com/mozilla-mobile/android-components/issues/9454 and know shape of the necessary Gecko delegates for credit cards, we can implement them.

TBD link to delegate work in bugzilla (per esawin).

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixes #9455 

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/GeckoCreditCardsAddressesStorageDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+) : CreditCardsAddressesStorageDelegate {
+
+    override fun onAddressFetch(): Deferred<List<Address>> {
+        return scope.async {
+            storage.value.getAllAddresses()
+        }
+    }
+
+    override fun onAddressSave(address: Address) {
+        TODO("Not yet implemented")
```
Comment (gabrielluong) [inline components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/GeckoCreditCardsAddressesStorageDelegate.kt:31]: I think `onAddressSave` and `onCreditCardSave` depends on our validation delegates which I have filed separately in https://github.com/mozilla-mobile/android-components/issues/9803

Comment (grigoryk) [reply components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/GeckoCreditCardsAddressesStorageDelegate.kt:31]: Just filed https://github.com/mozilla/application-services/issues/3899 - with that API in place, these calls can just go directly to storage.

E.g. at this point an address (and a credit card below) have been validated as part of the prompt flow, and we're just being asked by GV to save the data. Logins does validation here in order to figure out if it's an update or a save, but hopefully we won't need to do that.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt [hunk 2]
```diff
@@ -218,3 +220,32 @@ data class UpdatableAddressFields(
...
+    /**
+     * Saves the given address to storage.
+     */
+    fun onAddressSave(address: Address)
+
+    /**
+     * Returns all stored credit cards. This is called when the engine believes a credit card
+     * field should be autofilled.
+     */
+    fun onCreditCardFetch(): Deferred<List<CreditCard>>
```
Comment (grigoryk) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt:245]: see naming comments.

**Review by grigoryk — APPROVED** [review]
Naming nits, but it's good otherwise for now.

**Comment by grigoryk** [comment]
Just make sure we don't close #9455 until delegate is actually functional.

**Pull request merged**